File: src/server/logic/GameManager.java
package server.logic;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import shared.model.Player;
import shared.model.Role;

public class GameManager {
    public enum GamePhase { LOBBY, NIGHT, DAY, VOTING, END_GAME }

    private GamePhase currentPhase = GamePhase.LOBBY;
    private final Map<String, Player> players = new ConcurrentHashMap<>();
    private final DayActions dayActions = new DayActions();
    private final NightActions nightActions = new NightActions();

    public void addPlayer(String id, String name) {
        if (currentPhase == GamePhase.LOBBY) {
            players.put(id, new Player(id, name));
        }
    }

    public boolean isReadyToStart() {
        return players.size() >= 5 && currentPhase == GamePhase.LOBBY;
    }

    public void assignRoles() {
        if (!isReadyToStart()) return;

        List<Player> playerList = new ArrayList<>(players.values());
        Collections.shuffle(playerList);

        List<Role> roles = new ArrayList<>(Arrays.asList(
            Role.WEREWOLF, Role.SEER, Role.BODYGUARD, Role.VILLAGER, Role.VILLAGER
        ));
        
        while (roles.size() < playerList.size()) {
            roles.add(Role.VILLAGER);
        }
        Collections.shuffle(roles);

        for (int i = 0; i < playerList.size(); i++) {
            playerList.get(i).setRole(roles.get(i));
        }
        
        currentPhase = GamePhase.NIGHT;
    }

    public String advancePhase() {
        String resultMessage = "";
        switch (currentPhase) {
            case NIGHT -> {
                String nightVictim = nightActions.resolveNightDeath();
                if (nightVictim != null) {
                    Player p = players.get(nightVictim);
                    if (p != null) {
                        p.setAlive(false);
                        resultMessage = "Trời đã sáng! Đêm qua " + p.getName() + " đã bị sát hại.";
                    }
                } else {
                    resultMessage = "Trời đã sáng! Đêm qua là một đêm bình yên, không ai chết.";
                }
                currentPhase = GamePhase.DAY;
            }
                
            case DAY -> {
                currentPhase = GamePhase.VOTING;
                resultMessage = "Thời gian thảo luận đã hết. Bắt đầu bỏ phiếu!";
            }
                
            case VOTING -> {
                String lynchedPlayer = dayActions.getLynchedPlayer();
                if (lynchedPlayer != null) {
                    Player p = players.get(lynchedPlayer);
                    if (p != null) {
                        p.setAlive(false);
                        resultMessage = "Kết thúc bỏ phiếu. " + p.getName() + " đã bị treo cổ!";
                    }
                } else {
                    resultMessage = "Kết thúc bỏ phiếu. Không ai bị treo cổ vì hòa phiếu!";
                }
                
                nightActions.clearNightActions();
                dayActions.clearVotes();
                currentPhase = GamePhase.NIGHT;
                resultMessage += "\nTrời tối, mọi người đi ngủ!";
            }
                
            default -> {
            }
        }

        String winStatus = WinConditions.checkWin(new ArrayList<>(players.values()));
        if (!winStatus.equals("NO_WINNER_YET")) {
            currentPhase = GamePhase.END_GAME;
            resultMessage += "\nTRÒ CHƠI KẾT THÚC! " + winStatus;
        }

        return resultMessage;
    }

    public synchronized String handleDisconnect(String playerId) {
        Player p = players.get(playerId);
        if (p != null) {
            p.setConnected(false);
            p.setAlive(false);
            return WinConditions.checkWin(new ArrayList<>(players.values()));
        }
        return "NO_WINNER_YET";
    }

    public void resetGame() {
        players.clear();
        dayActions.clearVotes();
        nightActions.clearNightActions();
        currentPhase = GamePhase.LOBBY;
    }

    public GamePhase getCurrentPhase() { return currentPhase; }
    public void setCurrentPhase(GamePhase phase) { this.currentPhase = phase; }
    public Map<String, Player> getPlayers() { return players; }
    public DayActions getDayActions() { return dayActions; }
    public NightActions getNightActions() { return nightActions; }
}

File: src/server/logic/GameManager.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using WerewolfGame.Shared.Model;

namespace WerewolfGame.Server.Logic
{
    public enum GamePhase { LOBBY, NIGHT, DAY, VOTING, END_GAME }

    public class GameManager
    {
        private static readonly Random rng = new Random();

        public GamePhase CurrentPhase { get; set; } = GamePhase.LOBBY;
        public ConcurrentDictionary<string, Player> Players { get; } = new ConcurrentDictionary<string, Player>();
        public DayActions DayActions { get; } = new DayActions();
        public NightActions NightActions { get; } = new NightActions();

        public void AddPlayer(string id, string name)
        {
            if (CurrentPhase == GamePhase.LOBBY)
            {
                Players[id] = new Player(id, name);
            }
        }

        public bool IsReadyToStart()
        {
            return Players.Count >= 5 && CurrentPhase == GamePhase.LOBBY;
        }

        public void AssignRoles()
        {
            if (!IsReadyToStart()) return;

            List<Player> playerList = Players.Values.ToList();
            Shuffle(playerList);

            List<Role> roles = new List<Role>
            {
                Role.WEREWOLF, Role.SEER, Role.BODYGUARD, Role.VILLAGER, Role.VILLAGER
            };

            while (roles.Count < playerList.Count)
            {
                roles.Add(Role.VILLAGER);
            }
            Shuffle(roles);

            for (int i = 0; i < playerList.Count; i++)
            {
                playerList[i].Role = roles[i];
            }

            CurrentPhase = GamePhase.NIGHT;
        }

        public string AdvancePhase()
        {
            string resultMessage = "";

            switch (CurrentPhase)
            {
                case GamePhase.NIGHT:
                    // Đã thêm ? vào sau string
                    string? nightVictim = NightActions.ResolveNightDeath();
                    if (nightVictim != null)
                    {
                        if (Players.TryGetValue(nightVictim, out var p))
                        {
                            p.IsAlive = false;
                            resultMessage = "Trời đã sáng! Đêm qua " + p.Name + " đã bị sát hại.";
                        }
                    }
                    else
                    {
                        resultMessage = "Trời đã sáng! Đêm qua là một đêm bình yên, không ai chết.";
                    }
                    CurrentPhase = GamePhase.DAY;
                    break;

                case GamePhase.DAY:
                    CurrentPhase = GamePhase.VOTING;
                    resultMessage = "Thời gian thảo luận đã hết. Bắt đầu bỏ phiếu!";
                    break;

                case GamePhase.VOTING:
                    // Đã thêm ? vào sau string
                    string? lynchedPlayer = DayActions.GetLynchedPlayer();
                    if (lynchedPlayer != null)
                    {
                        if (Players.TryGetValue(lynchedPlayer, out var p))
                        {
                            p.IsAlive = false;
                            resultMessage = "Kết thúc bỏ phiếu. " + p.Name + " đã bị treo cổ!";
                        }
                    }
                    else
                    {
                        resultMessage = "Kết thúc bỏ phiếu. Không ai bị treo cổ vì hòa phiếu!";
                    }

                    NightActions.ClearNightActions();
                    DayActions.ClearVotes();
                    CurrentPhase = GamePhase.NIGHT;
                    resultMessage += "\nTrời tối, mọi người đi ngủ!";
                    break;
            }

            string winStatus = WinConditions.CheckWin(Players.Values);
            if (!winStatus.Equals("NO_WINNER_YET"))
            {
                CurrentPhase = GamePhase.END_GAME;
                resultMessage += "\nTRÒ CHƠI KẾT THÚC! " + winStatus;
            }

            return resultMessage;
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public string HandleDisconnect(string playerId)
        {
            if (Players.TryGetValue(playerId, out var p))
            {
                p.IsConnected = false;
                p.IsAlive = false;
                return WinConditions.CheckWin(Players.Values);
            }
            return "NO_WINNER_YET";
        }

        public void ResetGame()
        {
            Players.Clear();
            DayActions.ClearVotes();
            NightActions.ClearNightActions();
            CurrentPhase = GamePhase.LOBBY;
        }

        private static void Shuffle<T>(IList<T> list)
        {
            int n = list.Count;
            while (n > 1)
            {
                n--;
                int k = rng.Next(n + 1);
                T value = list[k];
                list[k] = list[n];
                list[n] = value;
            }
        }
    }
}